src/stdlib_stats_median.fypp

```diff
@@ -0,0 +1,255 @@
+#:include "common.fypp"
+#:set RANKS = range(1, MAXRANK + 1)
+#:set IR_KINDS_TYPES_OUTPUT = list(zip(INT_KINDS,INT_TYPES, ['dp']*len(INT_KINDS))) + list(zip(REAL_KINDS, REAL_TYPES, REAL_KINDS))
+
+submodule (stdlib_stats) stdlib_stats_median
+
+  use, intrinsic:: ieee_arithmetic, only: ieee_value, ieee_quiet_nan, ieee_is_nan
+  use stdlib_error, only: error_stop
+  use stdlib_optval, only: optval
+  ! Use "ord_sort" rather than "sort" because the former can be much faster for arrays
+  ! that are already partly sorted. While it is slightly slower for random arrays,
+  ! ord_sort seems a better overall choice.
+  use stdlib_sorting, only: sort => ord_sort
+  implicit none
+
+contains
+
+  #:for k1, t1, o1 in IR_KINDS_TYPES_OUTPUT
+    #:for rank in RANKS
+      #:set RName = rname("median_all",rank, t1, k1, o1)
+      module function ${RName}$ (x, mask) result(res)
+        ${t1}$, intent(in) :: x${ranksuffix(rank)}$
+        logical, intent(in), optional :: mask
+        real(${o1}$) :: res
+
+        integer(kind = int64) :: c, n
+        ${t1}$, allocatable :: x_tmp(:)
+
+        if (.not.optval(mask, .true.) .or. size(x) == 0) then
+          res = ieee_value(1._${o1}$, ieee_quiet_nan)
+          return
+        end if
+
+        #:if t1[0] == 'r'
+          if (any(ieee_is_nan(x))) then
+            res = ieee_value(1._${o1}$, ieee_quiet_nan)
+            return
+          end if
+        #:endif
+
+        n = size(x, kind=int64)
+        c = floor( (n + 1) / 2._${o1}$, kind=int64 )
+
+        x_tmp = reshape(x, [n])
```

Comment:
Is the reshape necessary here?

Reply:
Okay, I see this is to flatten the array. I guess `pack` could also be used in this case?

Reply:
The subroutine `sort` only accepts rank-1 array, while `median` support arrays of all ranks.

Do you have another suggestion to avoid the `reshape`?

Reply:
Indeed, `x_tmp = pack(x, .true.)` should work too. Do you think that `pack` is more efficient than `reshape` in this case?

Reply:
I imagine that a good compiler would do the same thing. So it can remain as is.